IL2CPP compliant build for Unity

jilleJr opened the pull request:
Adds IL2CPP specific constant checks in code according to @dngulin's guide at https://github.com/JamesNK/Newtonsoft.Json/issues/1440#issuecomment-430510160 via new build `net462-unity3d`

Manually tested in Unity 2018.3.11f1 in editor, WebGL, Windows (Mono), and Windows (IL2CPP).

Scraped together in a hurry. It compiles, tests passes for all builds including this one, deploys fine. Have tried following codebase structure as best I can. Please be brute' on things to fix for this PR and I will get on it.

dschofield93 commented:
Hey @jilleJr , great work on this feature - it's much needed for the project we're currently working on! Hey @JamesNK is there any update on whether this feature is planned to be merged into master? We're getting close to needing it for mobile deployment and it would be great to get the update through Nuget as opposed to managing a feature branch. Thanks!

jilleJr commented:
Yea well getting it in the actual releases would be grand. Maybe James isn't that active reading pull requests?

For those waiting but could do with building it manually, you can just clone my branch [https://github.com/jilleJr/Newtonsoft.Json#feature/unity-il2cpp-compiant](https://github.com/jilleJr/Newtonsoft.Json/tree/feature/unity-il2cpp-compiant) (same as from this PR) and start building the code locally by just executing the `Build/localbuild.ps1` powershell script.

```ps1
git clone https://github.com/jilleJr/Newtonsoft.Json --depth 1 --single-branch --branch feature/unity-il2cpp-compiant

# executed inside folder
Build/localbuild.ps1
```

Followed by just navigating to the folder `<repo>\Src\Newtonsoft.Json\bin\Release\net462-unity3d` and claim your dll's.

Admiral18 commented:
+1
We would need unity webGL support as well

jilleJr commented:
> @Admiral18
> +1
> We would need unity webGL support as well

This should in theory work for all IL2CPP configured builds, including **standalone (Windows/OS X/Linux), WebGL, iOS, Android, UWP,** etc. This new build configuration is good for all AOT compilations as well as some extra requirements needed for IL2CPP compilation.

So far tested and confirmed working on standalone & WebGL

ijovi23 commented:
> Yea well getting it in the actual releases would be grand. Maybe James isn't that active reading pull requests?
> 
> For those waiting but could do with building it manually, you can just clone my branch [https://github.com/jilleJr/Newtonsoft.Json#feature/unity-il2cpp-compiant](https://github.com/jilleJr/Newtonsoft.Json/tree/feature/unity-il2cpp-compiant) (same as from this PR) and start building the code locally by just executing the `Build/localbuild.ps1` powershell script.
> 
> ```powershell
> git clone https://github.com/jilleJr/Newtonsoft.Json --depth 1 --single-branch --branch feature/unity-il2cpp-compiant
> 
> # executed inside folder
> Build/localbuild.ps1
> ```
> 
> Followed by just navigating to the folder `<repo>\Src\Newtonsoft.Json\bin\Release\net462-unity3d` and claim your dll's.

Firstly thank you for your il2cpp version. However, when I try to build and attach to Unity in Visual Studio 2017, it failed. Did yours works well?

dbastien commented:
Can has please!

Spartanace1024 commented:
Would be great to have this pulled into master

jilleJr commented:
# Newtonsoft.Json for Unity is now available as an Unity Package!
## https://github.com/jilleJr/Newtonsoft.Json-for-Unity

For convenience I've prepared a Unity Package you all can abuse your socks off.

## Installation via Unity Package Manager (UPM)

> Requires Unity Editor `2018.1` or above, as UPM was first shipped with `2018.1`.

Open `<project>/Packages/manifest.json`, add the scope in `scopedRegistries` and the dependency `jillejr.newtonsoft.json-for-unity` like so:

```json
{
  "scopedRegistries": [
    {
      "name": "Packages from jillejr",
      "url": "https://registry.npmjs.org/",
      "scopes": ["jillejr"]
    }
  ],
  "dependencies": {
    "jillejr.newtonsoft.json-for-unity": "12.0.1",

    "com.unity.modules.ai": "1.0.0",
    "com.unity.modules.animation": "1.0.0",
    "com.unity.modules.assetbundle": "1.0.0",
    "com.unity.modules.audio": "1.0.0",
    "com.unity.modules.cloth": "1.0.0",
    "com.unity.modules.director": "1.0.0",
    "com.unity.modules.imageconversion": "1.0.0",
    "// cropping rest of packages for readability"
  }
}
```

Let's see how long I can keep this project up-to-date. I'm already behind as 12.0.2 just dropped.

## Note - far from alone

There are plenty of projects that gives JSON parsing in Unity. However most of them seems to be discontinued:

- `Json.NET v7` [nuget.org/packages/Unity.Newtonsoft.Json/](https://www.nuget.org/packages/Unity.Newtonsoft.Json/)
- `Json.NET v8` [parentelement.com/assets/json_net_unity](https://www.parentelement.com/assets/json_net_unity)
- `Json.NET v9` [github.com/SaladLab/Json.Net.Unity3D](https://github.com/SaladLab/Json.Net.Unity3D#readme)
- `Json.NET v12` [github.com/jilleJr/Newtonsoft.Json-for-Unity](https://github.com/jilleJr/Newtonsoft.Json-for-Unity#readme) _(ehy that's me!)_

While my repo will live for some time, for future reference here are some good alternatives _**(that all work in Unity)**_ that I personally recommend if my solution dips below your expectations:

- `OdinSerializer` [github.com/TeamSirenix/odin-serializer](https://github.com/TeamSirenix/odin-serializer#readme)
- `UTF8Json` [github.com/neuecc/Utf8Json](https://github.com/neuecc/Utf8Json#readme)
- `FastJSON` [assetstore.unity.com/packages/tools/input-management/fastjson-27220](https://assetstore.unity.com/packages/tools/input-management/fastjson-27220)
- `UltimateJson` [assetstore.unity.com/packages/tools/integration/ultimate-json-60845](https://assetstore.unity.com/packages/tools/integration/ultimate-json-60845)
- `SimpleJson` [wiki.unity3d.com/index.php/SimpleJSON](http://wiki.unity3d.com/index.php/SimpleJSON)
- `LitJson` [github.com/Mervill/UnityLitJson](https://github.com/Mervill/UnityLitJson#readme)
- `FullSerializer` [github.com/jacobdufault/fullserializer](https://github.com/jacobdufault/fullserializer#readme)
- `protobuf-net` [github.com/protobuf-net/protobuf-net](https://github.com/protobuf-net/protobuf-net#readme)

Happy Json parsing!

maskedmouse commented:
@jilleJr I've added the package to my new Unity project (2018.4.4 but also 2019.1.11) but building fails.

jilleJr commented:
> @jilleJr I've added the package to my new Unity project (2018.4.4 but also 2019.1.11) but building fails.
> Plugins are colliding for Windows x64 IL2CPP and manually fixing that, then build again it pretends the newtonsoft json namespace doesn't exist.

Super weird I did not find this before! Easily reproduced, even via local installation instead of UPM. It's like _"worked on my machine a week ago"_. Don't know what could be different now. 
I've added an issue in my repo with more details and I will tackle it as soon as I can. I'm interested in the what's and how's. Maybe Josh can help. Follow the issue over here: https://github.com/jilleJr/Newtonsoft.Json-for-Unity/issues/3

viktor-nedilko commented:
+1 It must have feature, Unity used wide novadays and has direct support will'b great

damdo commented:
Hey @JamesNK any hope to get this in the next release? It would be great to have Unity supported!

Dryra commented:
Hello @jilleJr,
Thank you for your fork, saved me a lot of time!
Just as info for someone trying this out and having problems with the script build.ps1, you have to change the path to MSBUILD.exe in GetMsBuildPath(), the path in the new Visual Studio 2019 is not "MSBuild\15.0\Bin\MSBuild.exe" rather MSBuild\Current\Bin\MSBuild.exe

So just change this line:
`return join-path $path 'MSBuild\15.0\Bin\MSBuild.exe'` 
with
`return join-path $path 'MSBuild\Current\Bin\MSBuild.exe'`

damdo commented:
Hey @JamesNK any hope to get this upstream? Having it support Unity would be really amazing for all of us! Thank you

AlexFolland commented:
I installed the above-linked UPM package from @jilleJr and switched my Unity 2019.3.4f1 project's Windows build scripting backend from Mono to IL2CPP.  My project was able to be built with Mono but not IL2CPP.  I got some obscure compiler error message about JsonConvert.DeserializeObject having an extra character that it didn't expect, but my JSON was completely valid.  From my research, I found and used the workaround in the following linked comment.  I just placed the code in the following linked comment directly before my JsonConvert.DeserializeObject call and it worked.

https://github.com/Azure/azure-iot-sdk-csharp/issues/213#issuecomment-341516646

I'm not sure why.  Maybe I am using an old version of the build chain or something.  Anyway, maybe this will help someone who had the same issue later.  Thanks to @jilleJr for this pull request and the Unity-compatible fork.  I hope @JamesNK can merge this pull request soon, as I don't see a downside.  Also thanks for the JSON tool.

jilleJr commented:
Hi @AlexFolland. This issue has had a lot of progress in the background by now. Here's the news:

- My fork is available via UPM and [OpenUPM](https://openupm.com) which makes it possible to use the DLLs without having to embed them into your repository and take up space there. I've not had good contact with James but my guess is that he doesn't have overly much time on adding that as a distribution point when there's alternatives available already.

- My fork has now a sister repo, the [jilleJr/Newtonsoft.Json-for-Unity.Converters](https://github.com/jilleJr/Newtonsoft.Json-for-Unity.Converters) where I've received help from both Dustin Horne (aka. ParentElement, author of the original [JSON .NET For Unity package from the Assets Store](https://assetstore.unity.com/packages/tools/input-management/json-net-for-unity-11347)) and Wanzyee Wang (aka. Wanzyee Studios, author of the most popular [Json.NET Converters package from the Assets Store](https://assetstore.unity.com/packages/tools/input-management/json-net-converters-simple-compatible-solution-58621)) to create an UPM package for delivering custom converters for the most common Unity types. A package that I guess neither Unity nor James will consider producing.

- Unity have gotten in contact and forked my repo and are now actually publishing some versions of my repo as the Newtonsoft.Json UPM package, officially built and verified by Unity, to their main UPM registry. Their package is already available from Unity v2020 but they plan to also make it available for lower version. Once their package is fully up and running I will consider just forwarding my package over to the Unity official package, but my Converters repo will continue living on.

There are some greater amount of work being put into making Newtonsoft.Json available to Unity (and particularly IL2CPP). I can only guess James' intention here being that he want to focus on supporting the .NET platform and let community around it make it available to other variants, as he seems to be putting less time into this project at the moment. Just merging this also makes him responsible for maintaining it and all of the sudden have to answer all Unity-specific questions. **These are just speculations** of course. Maybe @JamesNK can enlighten us with his intentions on this?
